crates/lsp/src/handlers_format.rs

```diff
@@ -41,8 +44,193 @@ pub(crate) fn document_formatting(
+    let root = air_r_factory::r_root(list, eof).build();
+
+    let format_info = biome_formatter::format_sub_tree(
+        root.syntax(),
+        air_r_formatter::RFormatLanguage::new(options),
+    )?;
+
+    if format_info.range().is_none() {
+        // Happens in edge cases when biome returns a `Printed::new_empty()`
+        return Ok(None);
+    };
+
+    let mut format_text = format_info.into_code();
+
+    // Remove last hard break line from our artifical expression list
+    format_text.pop();
+    let edits = to_proto::replace_range_edit(&doc.line_index, format_range, format_text)?;
+
+    Ok(Some(edits))
+}
+
+// From biome_formatter
+fn text_non_whitespace_range<E, L>(elem: &E) -> TextRange
+where
+    E: Into<SyntaxElement<L>> + Clone,
+    L: Language,
+{
+    let elem: SyntaxElement<L> = elem.clone().into();
+
+    let start = elem
+        .leading_trivia()
+        .into_iter()
+        .flat_map(|trivia| trivia.pieces())
+        .find_map(|piece| {
+            if piece.is_whitespace() || piece.is_newline() {
+                None
+            } else {
+                Some(piece.text_range().start())
+            }
+        })
+        .unwrap_or_else(|| elem.text_trimmed_range().start());
+
+    let end = elem
+        .trailing_trivia()
+        .into_iter()
+        .flat_map(|trivia| trivia.pieces().rev())
+        .find_map(|piece| {
+            if piece.is_whitespace() || piece.is_newline() {
+                None
+            } else {
+                Some(piece.text_range().end())
+            }
+        })
+        .unwrap_or_else(|| elem.text_trimmed_range().end());
+
+    TextRange::new(start, end)
+}
+
+/// Finds consecutive logical lines. Currently that's only expressions at
+/// top-level or in a braced list.
+fn find_deepest_enclosing_logical_lines(node: RSyntaxNode, range: TextRange) -> Vec<RSyntaxNode> {
+    let start_lists = find_expression_lists(&node, range.start(), false);
+    let end_lists = find_expression_lists(&node, range.end(), true);
+
+    // Both vectors of lists should have a common prefix, starting from the
+    // program's expression list. As soon as the lists diverge we stop.
+    let Some(list) = start_lists
+        .into_iter()
+        .zip(end_lists)
+        .take_while(|pair| pair.0 == pair.1)
+        .map(|pair| pair.0)
+        .last()
+    else {
+        // Should not happen as the range is always included in the program's expression list
+        tracing::warn!("Can't find common list parent");
+        return vec![];
+    };
+
+    let Some(list) = RExpressionList::cast(list) else {
+        tracing::warn!("Can't cast to expression list");
+        return vec![];
+    };
+
+    let iter = list.into_iter();
+
+    let logical_lines: Vec<RSyntaxNode> = iter
+        .map(|expr| expr.into_syntax())
+        .skip_while(|node| !node.text_range().contains(range.start()))
+        .take_while(|node| node.text_range().start() <= range.end())
+        .collect();
```

Comment:
In this case, were you expecting that it formats both `2 + 2` and THE ENTIRE `for` loop? i.e. note how `1 : 10` is formatted even though it is outside the selection.

It seems to be due to this bit here.

https://github.com/user-attachments/assets/adeb687f-201e-4730-be11-15bdab04e52a

Reply:
As I mentioned in https://github.com/posit-dev/air/pull/92#discussion_r1880952460 I was expecting that this _just_ formats `2 + 2` because that is the only node inside the common logical line "root" fully covered by the user's selection

Reply:
Should it be this?

```suggestion
    let logical_lines: Vec<RSyntaxNode> = iter
        .map(|expr| expr.into_syntax())
        .filter(|node| range.contains_range(node.text_trimmed_range()))
        .collect();
```

Careful to use `text_trimmed_range()` here, because we'd want this (where `<` and `>` represent the selection) to still format the `1+1` even though it doesn't span the full `text_range()` when you add in the comment

```
# hi there
<1+1>
```

This breaks one of your tests related to partial selection (reproduced in the image below) where I'd argue that only `1+1` should be reformatted

<img width="112" alt="Screenshot 2024-12-12 at 5 35 02 PM" src="https://github.com/user-attachments/assets/c40b095c-4368-49e4-8c1b-a2cadea3164a" />

Reply:
yes this was on purpose. We always format the selected part of user code, even if it means widening the range to the highest level of logical line selected by the user.

Reply:
I think what I don't love about this is that it means we can format lines arbitrarily far away from the user's actual selection

Like in this case I accidentally selected a little too far up and it means that the whole formatted is formatted, even if thats 20 lines further up


https://github.com/user-attachments/assets/0d1c9e3f-b3cc-4f33-9ded-c244374280d3

Reply:
We talked and decided to keep the current behavior, but with a comment around these lines and the fact that they allow the current implementation to expand the user's range, and add in my suggested code above as a commented out alternative that is more conservative if we ever wanted to switch to that